substrate/frame/scheduler/src/lib.rs

```diff
@@ -1192,6 +1311,92 @@ impl<T: Config> Pallet<T> {
+	///
+	/// If the task was successfully rescheduled for later, this function will return `Ok`.
+	/// Otherwise, if the task was not rescheduled, either because there was not enough weight left
+	/// over, there was no retry configuration in place, there were no more retry attempts left, or
+	/// the agenda was full, this function will return the task so it can be handled somewhere else.
+	fn try_schedule_retry(
+		weight: &mut WeightMeter,
+		now: BlockNumberFor<T>,
+		when: BlockNumberFor<T>,
+		agenda_index: u32,
+		task: ScheduledOf<T>,
+	) -> Result<(), ScheduledOf<T>> {
+		if let Err(()) =
+			weight.try_consume(T::WeightInfo::schedule_retry(T::MaxScheduledPerBlock::get()))
+		{
+			Self::deposit_event(Event::RetryFailed {
+				task: (when, agenda_index),
+				id: task.maybe_id,
+			});
+			return Err(task);
+		}
+
+		Self::schedule_retry(now, when, agenda_index, task)
+	}
+
+	/// Check if a task has a retry configuration in place and, if so, try to reschedule it.
+	///
+	/// If the task was successfully rescheduled for later, this function will return `Ok`.
+	/// Otherwise, if the task was not rescheduled, either because there was no retry configuration
+	/// in place, there were no more retry attempts left, or the agenda was full, this function
+	/// will return the task so it can be handled somewhere else.
+	fn schedule_retry(
+		now: BlockNumberFor<T>,
+		when: BlockNumberFor<T>,
+		agenda_index: u32,
+		task: ScheduledOf<T>,
+	) -> Result<(), ScheduledOf<T>> {
+		// Check if we have a retry configuration set.
+		match Retries::<T>::take((when, agenda_index)) {
+			Some(RetryConfig { total_retries, mut remaining, period }) => {
+				remaining = match remaining.checked_sub(1) {
+					Some(n) => n,
+					None => return Err(task),
+				};
+				let wake = now.saturating_add(period);
+				match Self::place_task(wake, task) {
+					Ok(address) => {
+						// Reinsert the retry config to the new address of the task after it was
+						// placed.
+						Retries::<T>::insert(
+							address,
+							RetryConfig { total_retries, remaining, period },
+						);
+						Ok(())
+					},
+					Err((_, task)) => {
+						// TODO: Leave task in storage somewhere for it to be
+						// rescheduled manually.
+						T::Preimages::drop(&task.call);
+						Self::deposit_event(Event::RetryFailed {
+							task: (when, agenda_index),
+							id: task.maybe_id,
+						});
+						Err(task)
+					},
+				}
+			},
+			None => Err(task),
+		}
+	}
+
+	/// Ensure that `left` has at least the same level of privilege or higher than `right`.
+	///
+	/// Returns an error if `left` has a lower level of privilege or the two cannot be compared.
+	fn ensure_privilege(
+		left: &<T as Config>::PalletsOrigin,
+		right: &<T as Config>::PalletsOrigin,
+	) -> Result<(), DispatchError> {
+		if matches!(T::OriginPrivilegeCmp::cmp_privilege(left, right), Some(Ordering::Less) | None)
+		{
+			return Err(BadOrigin.into());
+		}
+		Ok(())
+	}
```

Comment:
Is the `Retries` storage cleaned up when a task is cancelled?

Reply:
It wasn't, this just got lost in the impl, good catch! I also added a test for it.